File: 1356. Sort Integers by The Number of 1 Bits/sortByBits.js
/**
 * @param {number[]} arr
 * @return {number[]}
 */
var sortByBits = function (arr) {
    let ans = []
    let mp = new Map()

    for (let item of arr) {
        if (item === 0) {
            ans.push(0)
            continue
        }
        let bits = item.toString(2).match(/1/g)
        if (!mp.has(bits.length)) mp.set(bits.length, [])
        mp.get(bits.length).push(item)
    }
    const sortedMapByKey = new Map([...mp.entries()].sort((a, b) => a[0] - b[0]));

    // console.log("hash " , sortedMapByKey)

    for (let item of sortedMapByKey) {
        // console.log("item " , item[1])
        item[1].sort((a, b) => a - b)
        ans.push(...item[1])
    }
    // console.log("ans " , ans)
    return ans
};

File: 1356. Sort Integers by The Number of 1 Bits/sortByBits.cpp
#include <iostream>
#include <vector>
#include <map>
#include <bitset>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<int> sortByBits(vector<int>& arr) {
        vector<int> ans;
        map<int, vector<int>> mp;

        for(int num : arr){
            string binary = bitset<32>(num).to_string();
            int cnt = count(binary.begin(), binary.end(),'1');
            if(!mp.contains(cnt))mp.insert({cnt, {}});
            mp[cnt].push_back(num);
        }

        for(auto& pair : mp){
            vector<int>& vec = pair.second;
            sort(vec.begin(), vec.end());
            ans.insert(ans.end(),vec.begin(), vec.end());
        }

        return ans;
    }
};